physical/oci_objectstorage/oci_objectstorage.go

```diff
@@ -0,0 +1,398 @@
+	haEnabledStr := conf["ha_enabled"]
+	if haEnabledStr != "" {
+		haEnabled, err = strconv.ParseBool(haEnabledStr)
+		if err != nil {
+			return nil, errwrap.Wrapf("failed to parse HA enabled: {{err}}", err)
+		}
+
+		if haEnabled {
+			lockBucketName = conf["lockBucketName"]
+			if lockBucketName == "" {
+				return nil, errors.New("missing lock bucket name")
+			}
+		}
+	}
+
+	authTypeAPIKeyBool := false
+	authTypeAPIKeyStr := conf["authTypeAPIKey"]
+	if authTypeAPIKeyStr != "" {
+		authTypeAPIKeyBool, err = strconv.ParseBool(authTypeAPIKeyStr)
+		if err != nil {
+			return nil, errwrap.Wrapf("failed parsing authTypeAPIKey parameter: {{err}}", err)
+		}
+	}
+
+	var cp common.ConfigurationProvider
+	if authTypeAPIKeyBool {
+		cp = common.DefaultConfigProvider()
+	} else {
+		cp, err = auth.InstancePrincipalConfigurationProvider()
+		if err != nil {
+			return nil, errwrap.Wrapf("failed creating InstancePrincipalConfigurationProvider: {{err}}", err)
+		}
+	}
+
+	objectStorageClient, err := objectstorage.NewObjectStorageClientWithConfigurationProvider(cp)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed creating NewObjectStorageClientWithConfigurationProvider: {{err}}", err)
+	}
+
+	logger.Debug("physical/oci: configuration",
+		"bucketName", bucketName,
+		"namespaceName", namespaceName,
+		"haEnabled", haEnabled,
+		"lockBucketName", lockBucketName,
+		"authTypeAPIKey", authTypeAPIKeyBool,
+	)
+
+	return &Backend{
+		client:         &objectStorageClient,
+		bucketName:     bucketName,
+		logger:         logger,
+		permitPool:     physical.NewPermitPool(MaxNumberOfPermits),
+		namespaceName:  namespaceName,
+		haEnabled:      haEnabled,
+		lockBucketName: lockBucketName,
+	}, nil
+}
+
+func (o *Backend) Put(ctx context.Context, entry *physical.Entry) error {
+	o.logger.Debug("physical/oci: put")
+	defer metrics.MeasureSince(metricPutFull, time.Now())
+	startAcquirePool := time.Now()
+	metrics.SetGauge(metricPermitsUsed, float32(o.permitPool.CurrentPermits()))
+	o.permitPool.Acquire()
+	defer o.permitPool.Release()
+	metrics.MeasureSince(metricPutAcquirePool, startAcquirePool)
+
+	defer metrics.MeasureSince(metricPut, time.Now())
+	size := int64(binary.Size(entry.Value))
+	opcClientRequestId, err := uuid.GenerateUUID()
+	if err != nil {
+		o.logger.Debug("physical/oci: Put: error generating UUID")
+		opcClientRequestId = ""
+	}
+	o.logger.Debug("physical/oci: put", "opc-client-request-id", opcClientRequestId)
+	request := objectstorage.PutObjectRequest{
+		NamespaceName:      &o.namespaceName,
+		BucketName:         &o.bucketName,
+		ObjectName:         &entry.Key,
+		ContentLength:      &size,
+		PutObjectBody:      ioutil.NopCloser(bytes.NewReader(entry.Value)),
+		OpcMeta:            nil,
+		OpcClientRequestId: &opcClientRequestId,
+	}
+
+	resp, err := o.client.PutObject(ctx, request)
+	o.logRequest("put", resp.RawResponse, resp.OpcClientRequestId, resp.OpcRequestId, err)
+
+	if err != nil {
+		metrics.SetGauge(metricPutFailed, 1)
```

Comment:
Should these error gauges be counters instead? If they're intended to be gauges, shouldn't they be reset on success?

Reply:
This metric "metricPutFailed" emits 1 for any fail request and it will be absent if the PUT operation succeed. Gauge makes more sense to me since we don't do any counting here. What do you think?

Reply:
The last value of the gauge will implicitly be retained, so when reviewing your metrics you're going to probably see it presented as "1" at the first error and it will stay there indefinitely.  See the Counters section in the Prometheus docs, which references error counts:  https://prometheus.io/docs/concepts/metric_types/#counter

Reply:
Thanks for the detailed feedback! I switched counter for fail metrics.